pkg/cluster/kubernetes/talos_managed.go

```diff
@@ -288,3 +351,132 @@ func checkPodStatus(ctx context.Context, cluster UpgradeProvider, service, node,
 
 	return nil
 }
+
+//nolint:gocyclo,cyclop
+func checkDeprecated(ctx context.Context, cluster UpgradeProvider, options UpgradeOptions) error {
+	// disable kubernetes deprecations warnings
+	klog.LogToStderr(false)
+
+	options.Log("checking for resource APIs to be deprecated in version %s", options.ToVersion)
+
+	config, err := cluster.K8sRestConfig(ctx)
+	if err != nil {
+		return err
+	}
+
+	k8sClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		return fmt.Errorf("error building kubernetes client: %w", err)
+	}
+
+	staticClient, err := cluster.K8sHelper(ctx)
+	if err != nil {
+		return fmt.Errorf("error building kubernetes client: %s", err)
+	}
+
+	hasDeprecated := false
+
+	warnings := bytes.NewBuffer([]byte{})
+
+	w := tabwriter.NewWriter(warnings, 0, 0, 3, ' ', 0)
+
+	resources, ok := deprecations[options.Path()]
+	if !ok {
+		return nil
+	}
+
+	var namespaces *v1.NamespaceList
+
+	namespaces, err = staticClient.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	dc, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	serverResources, err := dc.ServerPreferredNamespacedResources()
+	if err != nil {
+		return err
+	}
+
+	namespacedResources := map[string]struct{}{}
+
+	for _, list := range serverResources {
+		for _, resource := range list.APIResources {
+			namespacedResources[resource.Name] = struct{}{}
+		}
+	}
+
+	for _, resource := range resources {
+		gvr, _ := schema.ParseResourceArg(resource)
+
+		if gvr == nil {
+			return fmt.Errorf("failed to parse group version resource %s", resource)
+		}
+
+		var res *unstructured.UnstructuredList
+
+		count := 0
+
+		probeResources := func(namespaces ...v1.Namespace) error {
+			r := k8sClient.Resource(*gvr)
```

Comment:
should we look in every namespace here then? 

I got why you had `*v1.Namespace` before, to pass `nil` there, I wonder if there's some better way? 

like pass empty list for cluster resources, and explicit list when it's namespaced? just throwing out ideas

Reply:
yeah. Passing list here is a nice change, will update :)